### PR TITLE
feat: new api endpoint for service updates

### LINKED ISSUE #982: Support editing of a Service
**Is your feature request related to a problem? Please describe.**
A user should be able to edit a service. In reality, at the moment, this would just cover the key-value data

**Describe the solution you'd like**
Support PATCH on `/api/v1/namespaces/<ns>/services/<service name>` 

**Additional context**
Related to https://github.com/epinio/epinio/issues/981, blocking UI issue https://github.com/epinio/ui/issues/9

### PR DESCRIPTION
**Pull request by andreas-kupries**
Fixes #982 

With model, backend, cli command

- manually tested basic functionality, without a bound app
- NOTE: restart through null update of deployment in doubt
- TODO: test suite expansion.

### CODE REVIEW

**Comment by andreas-kupries** [comment]
As thought, the null update does nothing, does not trigger a restart ... Good placeholder tough ;)

Searching

  - `kubectl rollout restart deployment <deployment name>` ... Now what action would that be in the API ...
  - `kubectl set env deployment <deployment name> DEPLOY_DATE="$(date)"` ... IOW set/change an environment variable ... That might be sensible if I cannot determine the rollout API.
  - `kubectl scale deployment <deployment name> --replicas=0` ... I.e. scale down and back up ... To be avoided, causes app down-time.

Aha ... https://stackoverflow.com/questions/59050709/how-to-rollout-restart-deployment-through-the-api ... The `rollout restart` simply patches an annotation in the pod template of the deployment ... Good, we know how to patch.

**Review by thardeck — APPROVED** [review]

File: pkg/api/core/v1/models/models.go [hunk 1]
```diff
@@ -38,6 +38,13 @@ type ServiceCreateRequest struct {
 	Data map[string]string `json:"data"`
 }
 
+// ServiceUpdateRequest represents and contains the data needed to
+// update a service instance (add/change, and remove keys)
+type ServiceUpdateRequest struct {
+	Remove []string          `json:"remove,omitempty"`
+	Edit   map[string]string `json:"edit,omitempty"`
```
Comment (richard-cox) [inline pkg/api/core/v1/models/models.go:45]: Would `Edit` be better described as `Set` given it covers adding new entries?

**Review by richard-cox — COMMENTED** [review]
Think most of the points I would raise have already been covered in https://github.com/epinio/epinio/issues/982#issuecomment-981446746 and https://github.com/epinio/epinio/issues/982#issuecomment-981597486. Going on the discussion, in an ideal world the UI would have a `put` which replaced the service data with the new set. As it stands a diff needs to occur locally in the browser/client side instead of the epinio backend. This is workable and removes the blocker but I think the logic should be occurring in the backend. If time/priority allows can the additional endpoint be added?

@andreas-kupries

File: pkg/api/core/v1/models/models.go (hunk 1 above)
Comment (andreas-kupries) [reply pkg/api/core/v1/models/models.go:45]: Ok. That is trivial.

**Comment by andreas-kupries** [comment]
>  This is workable and removes the blocker but I think the logic should be occurring in the backend. If time/priority allows can the additional endpoint be added?

Will you be ok if this PR is merged and the other endpoint comes in a second PR ? With BIT falling in my lap work on this has to be done around that, or by someone else on the team. Yet I also do not wish to keep you blocked.

I will do the `Edit`/`Set` change now, and resolve the conflict, i.e. rebase.

**Comment by andreas-kupries** [comment]
Rebase, and field rename: 6a714e5a3de265a53a9b56fcd6d8ef1b7c298c19

**Comment by andreas-kupries** [comment]
@jimmykarily @manno :warning:  The linter error for this last commit (rebased) is installer related, a checksum mismatch when trying to install tools. It did not even reach lintering the sources.

```
Installing epinio_installer
make: *** [Makefile:176: tools-install] Error 1
sha256 for https://github.com/epinio/installer/releases/download/v{version}/epinio-installer_linux_x86_64 does not match 27fc05bd820dd9790d0441a2b78b39c59fa717536eb710e627c55388fe39c4c7
Error: Process completed with exit code 2.
```

**Review by thardeck — APPROVED** [review]

**Pull request merged**